lib/utils/compose_ts.ts

```diff
@@ -1306,15 +1306,101 @@ async function getTokenForPreviousRepos(
 	return token;
 }
 
+async function pushAndUpdateServiceImages(
+	docker: Dockerode,
+	token: string,
+	images: TaggedImage[],
+	afterEach: (
+		serviceImage: import('balena-release/build/models').ImageModel,
+		props: object,
+	) => void,
+) {
+	const { DockerProgress } = await import('docker-progress');
+	const { retry } = await import('./helpers');
+	const { pushProgressRenderer } = await import('./compose');
+	const tty = (await import('./tty'))(process.stdout);
+	const opts = { authconfig: { registrytoken: token } };
+	const progress = new DockerProgress({ docker });
+	const renderer = pushProgressRenderer(
+		tty,
+		getChalk().blue('[Push]') + '    ',
+	);
+	const reporters = progress.aggregateProgress(images.length, renderer);
+
+	const pushImage = async (
+		localImage: Dockerode.Image,
+		index: number,
+	): Promise<string> => {
+		try {
+			// TODO 'localImage as any': find out exactly why tsc warns about
+			// 'name' that exists as a matter of fact, with a value similar to:
+			// "name": "registry2.balena-cloud.com/v2/aa27790dff571ec7d2b4fbcf3d4648d5:latest"
+			const imgName: string = (localImage as any).name || '';
+			const imageDigest: string = await retry({
+				func: () => progress.push(imgName, reporters[index], opts),
+				maxAttempts: 3, // try calling func 3 times (max)
+				label: imgName, // label for retry log messages
+				initialDelayMs: 2000, // wait 2 seconds before the 1st retry
+				backoffScaler: 1.4, // wait multiplier for each retry
+			});
+			if (!imageDigest) {
+				throw new ExpectedError(stripIndent`\
+					Unable to extract image digest (content hash) from image upload progress stream for image:
+					${imgName}`);
```

Comment:
Testing whether `imageDigest` is empty and, if so, throwing an error is the key change of this PR in relation to addressing the linked support thread. The rest of the PR is just conversion from Javascript to Typescript, and replacement of `Bluebird.using` with try/catch/finally. cc: @dfunckt @thgreasi (I have self certified, but post mortem feedback is still welcome!)